contributing/guide-observability.md

```diff
@@ -120,6 +120,61 @@ logs don't show up in production normally.
 
+## Good Errors
+
+An error message is good if it communicates
+
+1. what went wrong
+2. where it went wrong
+3. for which inputs
+
+and if it is concise without redundancies and at best displayable in one line.
+
+The general goal of an error message is to be actionable for the user who is trying
+to avoid the error in the future.
+
+Examples for bad error messages:
+- `cannot resolve package dependencies: Invalid Semantic Version`
+- `cannot parse file foo.yaml: failed to parse YAML of file foo.yaml`
+
+Examples for how to improve them:
+- `cannot resolve package dependencies in crossplane.yaml: invalid Semantic Version "main" on package "provider-kubernetes"`
+- `failed to load OCI image provider-gcp/v0.42.0: failed to unpack layer: cannot parse file: failed to parse YAML of file foo.yaml: unexpected indention in line 4`.
+
+General rules:
+1. follow the pattern `generic error cause: more details: even more details`.
+2. put values into quotes (`%q` in Golang), with the exception of
+   - resource names with namespace or resource type (e.g. `default/foo` or `providers/foo`), resource names alone are quoted
+   - kinds (e.g. `cannot find provider "provider-kubernetes"`; here, the kind provider is without quotes, the provider name is with)
+   - filenames with extensions and/or path (e.g. `cannot parse foo.yaml`).
+3. add a context of e.g. filenames, object names, or values which are wrong.
+
+   Which context is added by which functions depends on the level of abstraction of the code at hand: the package loader will know the OCI image name, the YAML parser might only see a byte slice. The former adds the image name as context, some layer in-between adds the filename and the YAML parser adds the line and column in the byte stream.
+
+   Filenames should be relative to the logical working directory of the context at hand (e.g. a package root, current working directory, workspace root). Ensure that the context is clear in the error message (e.g. `failed to load package provider-kubernetes: failed to parse apis/xrd.yaml`).
+
+6. add a context just once.
+
+   Rule of thumb: if a function gets a parameter, it will also include it in the error messages, so the caller does not have to.
+   
+   E.g. a function getting a filename as input will include the filename in errors it returns.
+7. don't return multi-line errors.
+8. aggregate errors if there is clear use for the user, but not in general, i.e. fail fast is the default.
+
+   Aggregate through `kerrors.NewAggregate`. When the number of aggregated errors is unbounded,
+   return a subset, e.g. the first 3.
+
+   E.g.: `cannot parse workspace: failed to parse foo.yaml, bar.yaml, abc.yaml, and 17 more files`
+9. don't include sensitive information like tokens into error messages, not even in debug mode.
+10. use error message constants, e.g. `errSomethingWentWrong` for messages without interpolation, and `errSomethingWentWrongFmt` for those with interpolation.
```

Comment:
The current pattern is to use `errFmt...`:
```suggestion
10. use error message constants, e.g. `errSomethingWentWrong` for messages without interpolation, and `errFmtSomethingWentWrong` for those with interpolation.
```

Reply:
we have both. Which do we want? Or both?

Reply:
Any thought in general on the usage of error constants?

My feeling is that it's encouraging some bad behaviours, such as:
- reusing errors from different sources, resulting in confusion on where an error came from, obviously the added context should make it unique, but it's painful to follow the whole chain, so ideally I would like errors to be kind of unique to be able to quickly find their source.
- brittle tests due to error's string representation comparison, I'd rather have structured errors if we actually want to check something specific.
- added cognitive load due to one more layer of indirection, in the end I always have to jump to the actual string to see what's the string, even more for `errFmt...` once, if I'm not the original author and I'm refactoring the code or even if I am, but I wrote it more than a week ago.

I know it's a well known pattern, so I'm not suggesting we change our approach, but given that we are discussing guidelines, I thought it could be worth asking it, worst case scenario in the end I'll have a better understanding of the benefits of this approach. 🙏

Reply:
given that we have 3 occurrences of `err...Fmt` and 76 of `errFmt...`, I'd say the former are mistakes and IIRC in a review by either @turkenh or @negz I was pointed to the latter as the standard indeed.

Reply:
+1 for getting rid of constants, for all the reasons you mention.

Reply:
Happy to unwind the error constant pattern, but let's tackle that in a separate PR. (I'm guessing that's already the intent since this thread is resolved.)